### DISCUSSION TITLE
기획 회의 일정

### DISCUSSION THREAD

**Discussion by jiyeon2** [wanted-pre-onboarding-side-project/habit-tracker#23] [Announcements]
## 일정

- 4월 10일 월요일 
- 커리어코칭 세션 완전히 끝난 후 또는 16시~19시

## 회의를 통해 확정 할 항목들

1. 필요 라이브러리
2. 상태 관리 어떻게 할까
3. 구현 기능 구체적으로 (참고 #1, #2)
4. 디자인

## 회의 전 준비 사항

:warning: 위 논의 항목들 4가지에 대한 의견과 근거 코멘트로 남기기

## 이미지 초안
![initiate](https://user-images.githubusercontent.com/67201870/230596146-ed934c99-3841-45a2-99e8-80d2b8009e0e.svg)

**Comment by jiyeon2** [comment]
오늘 디스코드 회의 때 디렉토리 구조 이야기도 나왔던거 같은데
'회의를 통해 확정 할 항목'에 디렉토리 구조도 포함하는 것은 어떨까요?

**Reply by greyHairChooseLife** [reply]
저는 반대합니다.

같은 날 회의할 내용들에 따라 디렉토리 구조도 많이 달라질텐데, 만약 이 회의 결과가 자기가 생각 해 본 것과는 많이 다르다면 며칠간 개발 하다가 '아 근데 디렉토리는 이렇게 하면 더 좋겠는데? 컴포넌트는 이렇게  나누면 좋겠는데?' 할 수 있을 것 같아요.

**Reply by greyHairChooseLife** [reply]
창환님, 따봉 이모티콘은 동의 한다는 뜻이신가요? 

지연님, 의견을 해 주신 부분에 대해서는 저도 따봉 드립니다. :+1:

**Reply by jiyeon2** [reply]
미리 디렉토리 구조를 협의하고 진행하면 
더 나은 구조가 떠올랐을때 기존에 협의한 구조에 얽매여 오히려 작업 생산성이 떨어질 우려가 있다는 말씀이 맞을까요?

디렉토리 구조는 코드 규모가 늘어나거나 하면 변경될 여지가 크니까
초반에 확정할 요소는 아닌거같네요!

다만 첫 에픽(기능)에 대한 작업을 베스트프랙티스로 선정하여 메인에 합친 이후
그걸 기반으로 다시 작업을 할 때
선정된 디렉토리 구조를 따른다든가 
그 구조를 기반으로 공통된 형태로 파일을 관리한다든가 하는 팀 내부 규칙은 있었으면 합니다~

**Reply by greyHairChooseLife** [reply]
@jiyeon2 

>더 나은 구조가 떠올랐을때 기존에 협의한 구조에 얽매여...

맞습니다. 

예를 들어 본인이 기획 회의 단계에서 redux를 쓰자고 제안했는데,  회의 결과 쓰지 않게 되었다면 '파일 및 디렉토리 구조'도 다시 생각 해 볼 여지가 많다는 것입니다. 회의 도중에 결정되는 라이브러리 사용 여부에 따라 **즉시** _"그러면 파일 구조는 이렇게 하면 좋겠네요" 혹은 "누구님 의견이 좋아 보겠습니다."_ 라고 의견 내기가 어려울 것 같습니다.

---

> 메인에 합친 이후 그걸 기반으로 다시 작업을 할 때, 선정된 디렉토리 구조를 따른다든가, 그 구조를 기반으로 공통된 형태로 파일을 관리한다든가 하는 팀 내부 규칙은 있었으면

이 부분 완전 동감합니다. 변경 할 분명한 이유와 근거가 뒷 받침 되지 않으면 기존 방식을 최대한 따르기로 정해두면 좋겠습니다.

또한 꼭 컨벤션으로 어딘가 써 두지 않아도, 이미 merge 된 코드에서 도출 될 수 있는 어떤 기준들은 가급적 지켜주면 좋을 것 같아요. 변경 시 `issue`나 `disccusion` 올리구요.

**Comment by jiyeon2** [comment]
### 필요 라이브러리

- react : 미사용시 ui처리 로직이 복잡해질 우려가 크다
- tailwindcss : 반응형 대응, 커스텀 스타일 적용 용이
  - chakra ui는 emotion에 의존성 있다. chakra ui를 프로젝트에 사용하면 emotion, chakraui 두 라이브러리 중 하나에만 문제가 생겨도 프로젝트에 영향을 받는다
- typescript : 타입체크로 런타임에서 발생할 가능성이 있는 에러를 컴파일 단계에서 해결 가능, interface와 type 작성으로 개발자들 간 소통 용이


### 상태관리
- react context api : 별도 상태관리 라이브러리 없어도 react에서 제공하는 api로 처리할 수 있을것으로 보임
  - 날짜
  - 해빗 목록
  - 생성창 여닫기

### 구현 기능 구체화 (이미지 초안으로 만든 ui 기준)
<ol>
<li><strong>날짜 표시 및 변경</strong>
<ul>
<li>기본값으로 ‘오늘’이 포함된 주를 표시
<ul>
<li>예 : 오늘이 4/9(일) 인 경우 4/3 ~ 4/9 가 표시됨</li>
</ul>
</li>
<li>버튼을 이용하여 다른 주로 변경 가능
<ul>
<li>현재 표시된 주가 ‘오늘’이 포함된 경우, 다음주로는 이동할 수 없다</li>
</ul>
</li>
<li>주의 첫 시작요일 : 월요일(일요일? 논의 필요)</li>
</ul>
</li>
<li><strong>주별 달성률 표시</strong>
<ul>
<li>주별 전체 해빗 달성률 = (현재 체크한 해빗 개수/ 이번주 전체 체크해야 할 해빗 개수 총 합)</li>
</ul>
</li>
<li><strong>해빗 목록 조회</strong>
<ul>
<li>최상단에 월, 화, 수, 목, 금, 토, 일 요일을 표시</li>
<li>생성버튼 표시</li>
<li>해빗 데이터가 많으면 스크롤바가 생긴다</li>
<li>실행할 해빗 항목 표시</li>
</ul>
</li>
<li><strong>해빗 표시</strong>
<ul>
<li>해당 해빗 이름 표시</li>
<li>습관 반복할 요일에 체크박스 표시 (특정 요일, 날짜에 실행여부 변경)</li>
<li>주간 달성 횟수 표시 = (이번 주 체크한 개수 / 이번 주 체크해야 할 개수)</li>
</ul>
</li>
<li><strong>해빗 생성, 수정, 삭제</strong>
<ul>
<li>
<p>서랍 띄우기, 숨기기</p>
</li>
<li>
<p>확인메시지 표시(삭제시)</p>
</li>
<li>
<p>기본값 및 유효성검사</p>


 항목 |  생성시 기본값 | 수정시 기본값 |  유효성검사 (만족해야 생성 가능)
-- | -- | -- | --
name | 빈문자열 | 클릭한 habit name | 최소 글자수 제한: 1, 최대 글자수 제한 : 20? 
| description | 빈문자열 | 클릭한 habit description | 최소 글자수 제한 : x, 최대 글자수 제한 : 100? 
| days | 7개 요일 전체 선택 | 클릭한 habit days | 최소 1개 선택 필수
</li>
</ul>
</li>

</ul>
</li>
<li><strong>해빗 목록 다중 선택 및 전체 삭제</strong>
<ul>
<li>전체 항목을 선택/ 일부 선택 및 해제 할 수 있다(전체삭제, 여러개 삭제 기능)</li>
<li>이미지 초안에는 버튼이 없지만, 전체 체크박스 위에 삭제버튼이 존재하는 형태 ⇒ 삭제확인메시지 표시 후 선택한 값 전체 삭제</li>
</ul>
</li>
</ol>
</li>
</ul>

**Comment by greyHairChooseLife** [comment]
## 1. 필요 라이브러리

- `chakra-ui` : 모든 멤버가 경험이 있는 styling UI 라이브러리

## 2. 상태 관리 어떻게 할까

```
interface Habit : {
    id: number
    name: string
    description: string
    days: Day[]   // type Day = '월' | '화' | ... | '일';
}
```
- `useContext`로 최상단에서 providing
- `useReducer`로 action type에 따라 create, update, delete 핸들러를 context provider에서 분리
- value와 handler를 각각 providing하여 효율적으로 리렌더

```
interface Record : {
    id: number
    startDate: Date
    days: Day[]
    achieve: weekRate[]    //  달성률 per week
}
```
- `useContext`로 최상단에서 providing
- (매 주 1회)**추가**만 하는 비교적 간단한 상태이므로 `useState` 사용
- 전체 기록 말고 당월 기록만 필요한 경우가 많으니 `return [당월, set당월, 전체]` 형태로 custom hook 만들어 사용
- value와 handler를 각각 providing하여 효율적으로 리렌더

## 3. 구현 기능 구체적으로 

1. habit create
    - 새로운 habit을 생성합니다.
    - 모든 요일이 활성화 된 상태를 기본값으로 합니다.

2. habit update
    - `days` 항목은 차주부터 적용됩니다.
    
4. habit delete
    - 개별 삭제만 구현(디자인 참고)

6. habit check
    - 습관 실행 여부를 당일(00:00~24:00)에만 표시할 수 있습니다.
    - 당일 체크 리스트가 **하이라이트** 됩니다.
    - 활성화 시간 내에 몇번이든 수정 할 수 있습니다.
    
7. record create
    - 지정한 요일 중 마지막 요일에 체크하면서 기록/수정 됩니다.

## 4. 디자인

    - 기존에 만든 것 모두 좋습니다!

**Comment by greyHairChooseLife** [comment]
회의 결과 보기 : [위키](https://github.com/wanted-pre-onboarding-side-project/habit-tracker/wiki/%ED%9A%8C%EC%9D%98%EB%A1%9D#2023-04-10-%EA%B8%B0%ED%9A%8D-%ED%9A%8C%EC%9D%98-discussion)
